pageserver/src/page_service.rs

```diff
@@ -200,6 +202,96 @@ impl PagestreamBeMessage {
     }
 }
 
+/// Implements Read for the server side of CopyIn
+struct CopyInReader<'a> {
+    pgb: &'a mut PostgresBackend,
+
+    /// Overflow buffer for bytes sent in CopyData messages
+    /// that the reader (caller of read) hasn't asked for yet.
+    /// TODO use BytesMut?
+    buf: Vec<u8>,
+
+    /// Bytes before `buf_begin` are considered as dropped.
+    /// This allows us to implement O(1) pop_front on Vec<u8>.
+    /// The Vec won't grow large because we only add to it
+    /// when it's empty.
+    buf_begin: usize,
```

Comment:
I hope there's a simpler solution.

The root cause of the problem is that I'm receiving data from a client that decides how much it wants to send, and giving it to the tar library, which decides how much it wants to read. So I have to buffer the bytes in between.

I tried using `VecDeque` as a buffer but it's missing lots of functionality, and I basically had to copy entries one by one in for loops. Not great.

This `Vec` version is fine, but would be nice if there was a simpler (less bug prone) existing solution, and maybe something that copies less. I was looking into `BytesMut`, I see we use it in `waldecoder`, but I'm not sure if `.advance()` does what I think it does.